### ISSUE TITLE
Root BO property - dependent upon property in all items of a child list

### ISSUE THREAD

**Issue by dazinator** [MarimerLLC/cslaforum#277]
Hello,
Is anyone able to enlighten me or point me to a sample showing the correct way to implement a parent that has a property, where that property is dependent upon properties within a child list item..

For example, an editable root called "Order" that as a child editable list of "OrderLine"'s - I'd like the root order BO to have an `OrderTotal` property which should basically be a calculated value, based on the Sum total of child `OrderLine`s (then minus order level discount etc)

My thinking is the order total property need to be a CSLA property info so that it can raise appropriate change notifications, and so that I can use the business rule system with it (I.E orders below 0 are invalid).

However the property needs to re-calculate each time a dependent property changes on a child item - and its that bit I am not sure what the correct way is to wire that up between the child and the parent.

I am wondering if I can wire this dependency up purely via the rules system, or if I have to hook into the property change events. I am also wondering if the parent should "subscribe" to the child items (so the children have no knowledge of the parent - which is my preference) or if the children need to "signal" the parent to recalc whenever particular properties on the child item are changed (not ideal from my perspective but if thats the only option I'll take it)

**Comment by rockfordlhotka** [comment]
There is a `ChildChanged` event that you can handle (or override `OnChildChanged`) in the root object, and in that event handler you can trigger the rules attached to your calculated property (and the rule would calculate the value). To execute rules for a specific property call `BusinessRules.CheckRules`.

The _Using CSLA 4: Creating Business Objects_ book has a direct example of doing what you are asking on page 91.

**Comment by jonnybee** [comment]
My recommendation is to override OnChildChanged event in the parent and call PropertyHasChanged for the "OrderTotal" property in order to trigger the rule for RecalculateOrderTotal. 

PropertyHasChanged will raise OnPropertyChanged (and trigger possibly dependent rules). If you Call BusinessRules.CheckRules then your code will also be reponsible for calling PropertyHasChanged.

**Comment by dazinator** [comment]
Thanks both! Excellent, that makes things much easier. I take it if the calculation logic is quite expensive, such that I only want to re-run it if certain properties on the child item change (rather than any property) - I take it the correct way to do that would be to query the event args in `OnChildChanged` and try to get infomation about the particular property on the child that was changed. I'll proceed along that route and thanks for the pointer!

**Comment by dazinator** [comment]
P.S I do own a few of your books Rocky, one of the first decent programming books I read on OO. I need to go back and re-read it I think, as it's been several years since my last CSLA despite me pushing it with every organisation I end up working for haha.

**Comment by dazinator** [comment]
Ok, so I tried.. but I think i must be missing something important.. as the `OnChildChanged` method on my parent / root BO doesn't appear to be triggered, when I change properties of an Item contained within a child list. I am wondering what I could have missed..

I'll include my code below just in case a glance at it reveals my ignorance (apologies for the volume, but I have tried to keep at minimal level).

Appreciate any hints!

This is what my parent / root BO looks like:

```
 public static readonly PropertyInfo<decimal> TotalNetProperty = RegisterProperty<decimal>(c => c.TotalNet);
        public decimal TotalNet
        {
            get { return GetProperty(TotalNetProperty); }
            set { SetProperty(TotalNetProperty, value); }
        }

 public static readonly PropertyInfo<OrderItemList> ItemsProperty = RegisterProperty<OrderItemList>(c => c.Items);
        public OrderItemList Items
        {
            get { return GetProperty(ItemsProperty); }
            set { SetProperty(ItemsProperty, value); }
        }

        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();

            BusinessRules.AddRule(new Lambda(TotalNetProperty, (ctx) =>
            {
                var sumOfItems = this.SupplierOrderItemList.Sum(a => a.TotalNet);
                // could minus any order level discount etc.              
                ctx.AddOutValue(sumOfItems);
            }));

        }

       // this doesn't get invoked..
        protected override void OnChildChanged(ChildChangedEventArgs e)
        {
            if (e.ChildObject is OrderItemList)
            {
                // will also cause business rules associated to the property to be re-run which in this
                // case, is the rule that re-calcualtes the order total. 
                PropertyHasChanged(TotalNetProperty);                
            }
            base.OnChildChanged(e);
        }

        public static Order Create()
        {
            return DataPortal.Create<Order>();
        }

        [RunLocal]
        protected override void DataPortal_Create()
        {
            LoadProperty<OrderItemList>(ItemsProperty, OrderItemList.CreateNewChildList());           
            base.DataPortal_Create();
        }

```

My child list:

```      

  [Serializable]
    public class OrderItemList : BusinessListBase<OrderItemList, IOrderItem>
    {
        internal static OrderItemList CreateNewChildList()
        {
            return DataPortal.CreateChild<OrderItemList>();
        }

        protected override void OnChildChanged(ChildChangedEventArgs e)
        {
           // this doesn't fire either..
            base.OnChildChanged(e);
        }

        protected override IOrderItem AddNewCore()
        {
            IOrderItem item = OrderItem.CreateChild();
            Items.Add(item);
            return item;          
        }
    }

```

And here is the item, and its interface hierarchy in case that's important:

```

 public interface IOrderItem : IBusinessBaseCore
    {
        int Id { get; }    

        decimal Price { get; set; }

        int Quantity { get; set; }

        decimal TotalNet { get; set; }
              
    }

    public interface IBusinessBaseCore : IBusinessBase
    {


    }

 [Serializable]
    public class OrderItem : BusinessBaseCore<OrderItem>, IOrderItem
    {

        public static readonly PropertyInfo<int> IdProperty = RegisterProperty<int>(c => c.Id);
        public int Id
        {
            get { return GetProperty(IdProperty); }
            private set { LoadProperty(IdProperty, value); }
        }        

        public static readonly PropertyInfo<decimal> PriceProperty = RegisterProperty<decimal>(c => c.Price);
        public decimal Price
        {
            get { return GetProperty(PriceProperty); }
            set { SetProperty(PriceProperty, value); }
        }

        public static readonly PropertyInfo<int> QuantityProperty = RegisterProperty<int>(c => c.Quantity);
        public int Quantity
        {
            get { return GetProperty(QuantityProperty); }
            set { SetProperty(QuantityProperty, value); }
        }           

        public static readonly PropertyInfo<decimal> TotalNetProperty = RegisterProperty<decimal>(c => c.TotalNet);
        public decimal TotalNet
        {
            get { return GetProperty(TotalNetProperty); }
            set { SetProperty(TotalNetProperty, value); }
        }       

        protected override void AddBusinessRules()
        {
            base.AddBusinessRules();
          
            // quantity must be 1 or higher
            BusinessRules.AddRule(new MinValue<int>(QuantityProperty, 1));

            // Set up properties so that TotalNet is automatically recalculated whenever Quantity, Price, are changed.
            BusinessRules.AddRule(new Dependency(PriceProperty, TotalNetProperty));
            BusinessRules.AddRule(new Dependency(QuantityProperty, TotalNetProperty));          

            BusinessRules.AddRule(new Lambda(TotalNetProperty, (ctx) =>
            {
               // This runs whenever quantity or price changes.
                var item = ctx.Target as OrderItem;              
                var sum = item.Quantity * item.Price;
                ctx.AddOutValue(sum);
            }));

            BusinessRules.AddRule(new MinValue<decimal>(TotalNetProperty, 0m)); // net total shouldn't fall below 0.

        }


        #region Factory Methods

        internal static OrderItem CreateChild()
        {
            var item = DataPortal.CreateChild<OrderItem>();
            return item;
        }

        #endregion
       
    }

```


I have a unit test in which i am doing:

```

 var sut = Order.Create();
 var item = sut.SupplierOrderItemList.AddNew();
 item.Quantity = 1;
 Assert.True(item.IsValid);
 Assert.Equal(0, item.TotalNet);

 item.Price = 10;
 Assert.Equal(10m, item.TotalNet);
 Assert.True(item.IsValid);

 item.Quantity = 2;
 Assert.Equal(20m, item.TotalNet);

  // order total should be up to date.
var orderTotal = sut.TotalNet;
var expectedTotal = item.TotalNet;
Assert.Equal(expectedTotal, orderTotal); / FAILS

```

At the item level, the calculated total of each item is working as desired.

However as I am changing properties of each item (like quantity, price) OnChildChanged is not invoked.
I am wondering whether I need to be doing something additional to hook this up

**Comment by ck-g** [comment]
I suppose it is the OnChildChanged method in the parent where you ask about the child list but not the child object itself.

Please try:
if (e.ChildObject is OrderItem)
{ }

Instead of: 
if (e.ChildObject is OrderItemList)
{ }

**Comment by dazinator** [comment]
@ck-g - the entire method doesn't even enter. So in other words, it's not even getting to that if() statement..

**Comment by dazinator** [comment]
Haven't been able to solve this.. So I am going to create a branch and will try and replicate this issue with a unit test.

**Comment by dazinator** [comment]
Ok I have got to the bottom of this problem! The existing unit tests that CSLA has around the collection classes run fine. But none of the lists involved seem to override AddNewCore().. The problem I am seeing was due to my override of `AddNewCore` on my List.

If you don't call `base.AddNewCore()` then child changed events won't be propagate.

Here was my original implementation:

```

 protected override OrderItem AddNewCore()
        {           
            OrderItem item = OrderItem.CreateChild();
            Items.Add(item);
            return item;
        }
```

I think it used to be in days of old that the base class implementation of this method would throw an exception if you didn't override it.. So you wouldn't call into it, you'd just override it. This is probably still the case in `BusinessBindingListBase` not sure.

However with BusinessListBase it looks like the correct implementation is this:

```
      protected override OrderItem AddNewCore()
        {
            var item = base.AddNewCore();
            return item;       
        } 

```

In other words it doesn't look like you need to override this method anymore :-)

**Comment by dazinator** [comment]
Nope I'm a lemon.. You do have to override AddNewCore still if the child item doesn't have a default constructor. 

So the implementation I was actually looking for was one line away from my original..
rather than Items.Add(), I just wanted `Add()` !

```
 protected override SupplierOrderItem AddNewCore()
        {          
            var item = SupplierOrderItem.CreateChild();
            Add(item);
           // Items.Add(item); // change notifications won't propagate if you do this! 
            return item;
        }

```

**Comment by dazinator** [comment]
For future reference, this is what `OnChildChanged` ended up looking like on my parent class, 
in order to only re-calc the order total, when related events / properties are changed:

```
     protected override void OnChildChanged(ChildChangedEventArgs e)
        {

            bool recalc = false;

            if (e.ChildObject is OrderItemList)
            {
                if (e.CollectionChangedArgs != null)
                {
                    recalc = true; // the list is having items added or removed..
                }             
            }
            else if (e.ChildObject is ISupplierOrderItem) // An item in the list has been edited..
            {
                if (e.PropertyChangedArgs != null)
                {
                    switch (e.PropertyChangedArgs.PropertyName)
                    {
                        case nameof(ISupplierOrderItem.TotalNet): // the line total of the item has changed                   
                            recalc = true;
                            break;
                    }
                };
            }

            if (recalc)
            {
                // trigger total net to recalc.
                PropertyHasChanged(TotalNetProperty); // causes rule to be tigerred against this property which will recalc the order total.
            }

            base.OnChildChanged(e);
        }

```

**Comment by dazinator** [comment]
The above code isn't quite as nice as the code we can use in the RuleSystem for wiring up dependency properties. Its a lot of manual checking of the event args and objects raised via events.

If anyone is aware of a way to achieve the above but a bit cleaner I'd be interested to see.

**Comment by jonnybee** [comment]
Do you experience any lag in the UI if you just simplify the code to: 

     protected override void OnChildChanged(ChildChangedEventArgs e)
        {
            PropertyHasChanged(TotalNetProperty); 
            base.OnChildChanged(e);
        }

BusinessRules is run quite often and this is just a summary in memory and should be very fast.

**Comment by dazinator** [comment]
@jonnybee - yeah I take your point, and ordinarily I'd be fine with that if this was a windows forms application.. (or wpf) but i'm working on a web application (asp.net core), and I am expecting orders to be synchronised from multiple users devices to the website. Each device basically sends json to a web method where parse it and create the equivalent csla Order business objects saving them to the database. Performing unnecessary evalutations of rules in this context does start to make a difference when you scale it by volume of users and data, and I also expect the order calculation logic to get a little more complicated over time for example - I expect discounts to make an appearance at some point, both at the line item level and the order level! Lucky me

**Comment by jonnybee** [comment]
Do you use the CslaModelBinder or the default ModelBinder i ASP.NET MVC?  

In a web app you would typically suppress Rule execution during ModelBinding and then call BusinessRules.CheckRules after binding is completed.

**Comment by dazinator** [comment]
> Do you use the CslaModelBinder or the default ModelBinder i ASP.NET MVC?

I don't believe the `CslaModelBinder` is available for use with asp.net core yet - although I could be wrong?

I'm using MVC model binding to bind to DTO's which are accepted into my web service method, based on the json data sent from the client (an xamarin app). I then iterate these DTO's constructing and saving proper CSLA business objects. 

> In a web app you would typically suppress Rule execution during ModelBinding and then call BusinessRules.CheckRules after binding is completed.

Ah ok that makes sense! So in my scenario then I should probably: `order.SuppressRuleChecking` whilst I am constructing the order root BO, and then `sut.ResumeRuleChecking` once its fully populated and ready to save?

**Comment by jonnybee** [comment]
You can access these method by the ICheckRules Interface on any BusinessBase Object: 

    public interface ICheckRules
    {
        void SuppressRuleChecking();
        void ResumeRuleChecking();
        void CheckRules();
        BrokenRulesCollection GetBrokenRules();
    }

Which allows your code to determine when rules are executed.

**Comment by dazinator** [comment]
> ReumeRuleChecking just resets the flag. You must Call CheckRules to run all rules.

Ok excellent - thank you!
